internals/overlord/checkstate/handlers.go

```diff
@@ -0,0 +1,198 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package checkstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	tombpkg "gopkg.in/tomb.v2"
+
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/overlord/state"
+	"github.com/canonical/pebble/internals/plan"
+)
+
+func (m *CheckManager) doPerformCheck(task *state.Task, tomb *tombpkg.Tomb) error {
+	m.state.Lock()
+	changeID := task.Change().ID()
+	var details checkDetails
+	err := task.Get(checkDetailsAttr, &details)
+	config := m.state.Cached(performConfigKey{changeID}).(*plan.Check)
+	m.state.Unlock()
+	if err != nil {
+		return fmt.Errorf("cannot get check details for perform-check task %q: %v", task.ID(), err)
+	}
+
+	logger.Debugf("Performing check %q with period %v", details.Name, config.Period.Value)
+	ticker := time.NewTicker(config.Period.Value)
+	defer ticker.Stop()
+
+	chk := newChecker(config)
+	for {
+		select {
+		case <-ticker.C:
+			err := runCheck(tomb.Context(nil), chk, config.Timeout.Value)
+			if err != nil {
+				// Record check failure and perform any action if the threshold
+				// is reached (for example, restarting a service).
+				details.Failures++
+				atThreshold := details.Failures >= config.Threshold
+				if !atThreshold {
+					// Update number of failures in check info. In threshold
+					// case, check info will be updated with new change ID by
+					// changeStatusChanged.
+					m.updateCheckInfo(config, changeID, details.Failures)
+				}
+
+				m.state.Lock()
+				if atThreshold {
+					details.Proceed = true
+				} else {
+					// Add error to task log, but only if we haven't reached the
+					// threshold. When we hit the threshold, the "return err"
+					// below will cause the error to be logged.
+					logTaskError(task, err)
+				}
+				task.Set(checkDetailsAttr, &details)
+				m.state.Unlock()
+
+				logger.Noticef("Check %q failure %d/%d: %v", config.Name, details.Failures, config.Threshold, err)
+				if atThreshold {
+					logger.Noticef("Check %q threshold %d hit, triggering action and recovering", config.Name, config.Threshold)
+					m.callFailureHandlers(config.Name)
+					// Returning the error means perform-check goes to Error status
+					// and logs the error to the task log.
+					return err
+				}
+			} else if details.Failures > 0 {
+				m.updateCheckInfo(config, changeID, 0)
+
+				m.state.Lock()
+				task.Logf("succeeded after %s", pluralise(details.Failures, "failure", "failures"))
+				details.Failures = 0
+				task.Set(checkDetailsAttr, &details)
+				m.state.Unlock()
+			}
+
+		case <-tomb.Dying():
+			break
+		}
+
+		// Do this check here so that select doesn't sometimes pick up another
+		// ticker.C tick after the tomb has been killed.
+		tombErr := tomb.Err()
```

Comment:
This doesn't actually protect you against this, just mitigates it somewhat.

Reply:
Ah yes. I've rearranged this per our discussion and your comments below.